kratos/geometries/nurbs_volume_geometry.h

```diff
@@ -0,0 +1,997 @@
+        }
+
+        for (IndexType shape_function_row_i = 0;
+            shape_function_row_i < shape_function_container.NumberOfShapeFunctionRows();
+            ++shape_function_row_i) {
+            for (IndexType u = 0; u <= PolynomialDegreeU(); ++u) {
+                for (IndexType v = 0; v <= PolynomialDegreeV(); ++v) {
+                    for (IndexType w = 0; w <= PolynomialDegreeW(); ++w) {
+                        IndexType cp_index_u = shape_function_container.GetFirstNonzeroControlPointU() + u;
+                        IndexType cp_index_v = shape_function_container.GetFirstNonzeroControlPointV() + v;
+                        IndexType cp_index_w = shape_function_container.GetFirstNonzeroControlPointW() + w;
+
+                        const IndexType index = NurbsUtilities::GetVectorIndexFromMatrixIndices(
+                            NumberOfControlPointsU(), NumberOfControlPointsV(), NumberOfControlPointsW(),
+                                cp_index_u, cp_index_v, cp_index_w);
+
+                        if (u == 0 && v==0 && w==0)
+                            rGlobalSpaceDerivatives[shape_function_row_i] =
+                            (*this)[index] * shape_function_container(u, v, w, shape_function_row_i);
+                        else
+                            rGlobalSpaceDerivatives[shape_function_row_i] +=
+                            (*this)[index] * shape_function_container(u, v, w, shape_function_row_i);
+                    }
+                }
+            }
+        }
+    }
+
+    ///@}
+    ///@name Shape Function
+    ///@{
+
+    Vector& ShapeFunctionsValues(
+        Vector &rResult,
+        const CoordinatesArrayType& rCoordinates) const override
+    {
+        NurbsVolumeShapeFunction shape_function_container(mPolynomialDegreeU, mPolynomialDegreeV, mPolynomialDegreeW, 0);
+
+        if (IsRational()) {
+            shape_function_container.ComputeNurbsShapeFunctionValues(mKnotsU, mKnotsV, mKnotsW, mWeights,
+                rCoordinates[0], rCoordinates[1], rCoordinates[2]);
+        }
+        else {
+            shape_function_container.ComputeBSplineShapeFunctionValues(mKnotsU, mKnotsV, mKnotsW,
+                rCoordinates[0], rCoordinates[1], rCoordinates[2]);
+        }
+
+        if (rResult.size() != shape_function_container.NumberOfNonzeroControlPoints())
+            rResult.resize(shape_function_container.NumberOfNonzeroControlPoints());
+
+        for (IndexType i = 0; i < shape_function_container.NumberOfNonzeroControlPoints(); ++i) {
+            rResult[i] = shape_function_container(i, 0);
+        }
+
+        return rResult;
+    }
+
+    Matrix& ShapeFunctionsLocalGradients(
+        Matrix& rResult,
+        const CoordinatesArrayType& rCoordinates) const override
+    {
+        NurbsVolumeShapeFunction shape_function_container(mPolynomialDegreeU, mPolynomialDegreeV, mPolynomialDegreeW, 1);
```

Comment:
Inside the nurbs_surface_geometry.cpp the last argument is 0. So no derivatives will be computed. That does not make sense right? It should be one?

Reply:
ok, that should be an error then. That's correct. because it indicates how much memory is kept for the derivatives...

Reply:
I can put the updates of the surface in the separate PR?